Documentation Updates for Helm Migration (#206)
redhat-cop/rhel-edge-automation-arch: README.md

![Overall Architecture](/images/overall-architecture.png)

> ## Above Site Components

Author: I removed a section in the first paragraph about this repository being a work in progress. Should I add that back in but change the wording to show the above site has matured and focus will shift to below site in future releases?

> ## Above Site Components

Reviewer: lets remove the work in progress. we have a solid above site implementation that can be run as is at this time.

We can leave the below site assets until we get there.

OpenShift can be used to host all of the above site components. These components include:

* Helm/Argo CD for deployment
* OpenShift Virtualization for RHEL Image Builder
* OpenShift Pipelines driving Ansible playbooks
* Nexus for artifact storage
* ODF (Noobaa only) for general object storage
* Quay to host RFE OSTree content

## Deploying Above Site Components

[Helm](https://helm.sh) and [Argo CD](https://argoproj.github.io/argo-cd/) are used to deploy and manage project components. Helm is used to dynamically generate an app of apps pattern in Argo CD, which in turn will pull in all the necessary Helm charts to deploy the specific components needed in the target environment.

> Before beginning, make sure you have the latest versions of `oc`/`kubectl`, `git` and `helm` clients installed.

Author: ACK

![Overall Architecture](/images/overall-architecture.png)

## Above Site Components

> OpenShift can be used to host all of the above site components. These components include:

Reviewer 2: Nitpick: I'd use stronger words here since we're very much dependent on OCP and RHEL.

Author: I can reword that. The intent was to show that you can BYO some of the components but that is not clear here.

* Helm/Argo CD for GitOps based deployment and configuration
* OpenShift Virtualization for RHEL Image Builder
* OpenShift Pipelines driving Ansible playbooks
* Nexus for artifact storage
* OpenShift Data Foundation (NooBaa only) for general object storage
* Red Hat Quay to host RFE OSTree content

## Deploying Above Site Components

[Helm](https://helm.sh) and [Argo CD](https://argoproj.github.io/argo-cd/) are used to deploy and manage project components. Helm is used to dynamically generate an app of apps pattern in Argo CD, which in turn will pull in all the necessary Helm charts to deploy the specific components needed in the target environment.

> Before beginning, make sure you have the latest versions of `oc`/`kubectl`, `git` and `helm` clients installed.

Reviewer 2: Do we need to mention `tkn` too?
https://tekton.dev/docs/cli/

Author: Yeah I'll add that.

```shell
ln -s ~/.ssh/image-builder charts/bootstrap/files/ssh/image-builder-ssh-private-key
ln -s ~/.ssh/image-builder.pub charts/bootstrap/files/ssh/image-builder-ssh-public-key
```

The rest of the values will be defined in a Helm values file. In the root of the repository, create a file called `local/bootstrap.yaml` and add the following:

```yaml
rhsm:
  portal:
    secretName: redhat-portal-credentials
    offlineToken: "Opij2qw3eCf890ujjwec8j..."
    password: "changeme"
    poolId: "ssa77eke7ahs0123djsdf92340p9okjd"
    username: "alice"

global:
  git:
    url: https://github.com/redhat-cop/rhel-edge-automation-arch.git
    ref: main
```

Be sure to change the values of `offlineToken`, `poolId`, `username`, and `password` to match the details for your account. If you are not sure how to generate an offline token for the Red Hat API, it is documented [here](https://access.redhat.com/articles/3626371#bgenerating-a-new-offline-tokenb-3).

#### Deploy OpenShift GitOps Operator and Argo CD

Once the SSH keypair and values file are in place, we can begin to deploy. Run the following script to install the OpenShift GitOps Operator and Argo CD.

```shell
./setup/init.sh
```

> 

Reviewer 2: Should we mention to wait until the `rfe-gitops` namespace has finished reconciliation before progressing to the next step? Helm will fail to install the Application manifests until the CRD coming from the ArgoCD subscription has been created in the cluster. WDYT?

Reviewer: Doesn't [this](https://github.com/redhat-cop/rhel-edge-automation-arch/blob/helm-migration/setup/init.sh#L12) accomplish what you are asking for?

```

#### Customizing Components

Each chart in the `charts/` directory has a default values file. These values can be overwritten using the same pattern shown above in _Disabling Components_.

For example, to enable processor emulation for OpenShift Virtualization, set `useEmulation: true` in the chart's values file. Store the following file in `local/cnv-processor-emulation.yaml`:

```yaml
---
# Dynamically Generated Charts
application-manager:
  charts:
    # Top Level RFE App of App Chart
    rfe-automation:
      values:
        charts:
          # Cluster Configuration App of App Chart
          cluster-configs:
            values:
              charts:
                # OpenShift Virtualization
                cnv:
                  values:
                    cnv:
                      debug:
                        useEmulation: "true"
```

Pass this values file to helm when deploying the project. For example:

```shell
helm upgrade -i -n rfe-gitops bootstrap charts/bootstrap/ -f local/bootstrap.yaml -f examples/helm/deployment/default.yaml -f local/cnv-processor-emulation.yaml
```

> 

Reviewer 2: What about adding at the end of the installation a screenshot of the ArgoCD apps synchronized as an example of the expected outcome from the helm upgrade command?

Reviewer 2: Or this perhaps:
```bash
$> oc get app -n rfe-gitops
NAME                       SYNC STATUS   HEALTH STATUS
ansible-rfe-runner         Synced        Healthy
catalog-sources            Synced        Healthy
cluster-configs            Synced        Healthy
cnv                        Synced        Healthy
cnv-operator               Synced        Healthy
httpd                      Synced        Healthy
image-builder-vm           Synced        Healthy
namespaces                 Synced        Healthy
nexus                      Synced        Healthy
nexus-operator             Synced        Healthy
operators                  Synced        Healthy
pipelines-operator         Synced        Healthy
quay                       Synced        Healthy
quay-operator              Synced        Healthy
rbac                       Synced        Healthy
resource-locker-operator   Synced        Healthy
rfe                        Synced        Healthy
rfe-automation             Synced        Healthy
rfe-pipelines              Synced        Healthy
user-mgmt                  Synced        Healthy
user-workload-monitoring   Synced        Healthy
```

Reviewer: If we change the name of any of the apps in the future or add/remove, it will quickly get out of sync

Author: How about we meet in the middle? Add a step to validate all applications in the `rfe-gitops` namespace should be Synced/Healthy?

Reviewer 2: Yep, that should do it.

Reviewer 2: You could even go further and just check for the status of the ref-automation app only... since they all derive from it.

> | Pool ID                      | Pool ID use to map the appropriate subscription to the Image Builder VM |

Reviewer: ```suggestion
| Pool ID                      | Red Hat Subscription Manager Pool ID use to map the appropriate subscription to the Image Builder VM |
```

Reviewer: Should we also look at needing to support RH Satellite in the future?

Author: Yeah, we have an old issue (#63) for Image Builder tracking this. In the before times there were issues getting Satellite backed repos working with Image Builder. I'm fairly certain those have all been resolved now, so we need to dig that back up.

| Red Hat Portal Offline Token | Token used to access the Red Hat API and download RHEL images           |

To generate the SSH keypair, run the following command:

```shell
ssh-keygen -t rsa -b 4096 -C cloud-user@image-builder -f ~/.ssh/image-builder
```

From the root of the repository, create symlinks to the key pair you just created:

```shell
ln -s ~/.ssh/image-builder charts/bootstrap/files/ssh/image-builder-ssh-private-key
ln -s ~/.ssh/image-builder.pub charts/bootstrap/files/ssh/image-builder-ssh-public-key
```

> The rest of the values will be defined in a Helm values file. In the root of the repository, create a file called `local/bootstrap.yaml` and add the following:

Reviewer: Potential thoughts for where to place values files. Maybe have top level folder called `values` and then subdirectories of `local` and `examples`?

Reviewer 2: s/values/config?

Author: +1 on values/{local,examples}

Author: As discussed settling on `examples/values` pattern.

```yaml
rhsm:
  portal:
    secretName: redhat-portal-credentials
    offlineToken: "Opij2qw3eCf890ujjwec8j..."
    password: "changeme"
    poolId: "ssa77eke7ahs0123djsdf92340p9okjd"
    username: "alice"

> global:

Reviewer: We can omit the use of the `global` values

  git:
    url: https://github.com/redhat-cop/rhel-edge-automation-arch.git
    ref: main
```

> Be sure to change the values of `offlineToken`, `poolId`, `username`, and `password`. If you are not sure how to generate an offline token for the Red Hat API, it is documented [here](https://access.redhat.com/articles/3626371#bgenerating-a-new-offline-tokenb-3).

Reviewer: ```suggestion
Be sure to change the values of `offlineToken`, `poolId`, `username`, and `password` to match the details for your account. If you are not sure how to generate an offline token for the Red Hat API, it is documented [here](https://access.redhat.com/articles/3626371#bgenerating-a-new-offline-tokenb-3).
```

#### Deploy OpenShift GitOps Operator and Argo CD

Once the SSH keypair and values file are in place, we can begin to deploy. Run the following script to install the OpenShift GitOps Operator and Argo CD.

```shell
./setup/init.sh
```

### Deploying

To deploy a reference environment in an empty OpenShift cluster, run the following command:

```shell
helm upgrade -i -n rfe-gitops bootstrap charts/bootstrap/ -f local/bootstrap.yaml -f examples/helm/deployment/default.yaml
```

The default installation will deploy and configure all of the managed components on the cluster. An HTPasswd identity provider is configured for 5 users (`user{1-5}`) with `openshift` as the password.

You can track the progress of the deployment on the Argo CD dashboard. To get the URL run the following command:

```shell
oc get route argocd-server -n rfe-gitops -ojsonpath='https://{.spec.host}'
```

The parent application is `rfe-automation`.

### Customizing the Deployment

> Helm and Argo CD are used to deploy and manage all of the project components. From a high level, a Helm chart called `application-manager` is used to dynamically build nested app of apps pattern in Argo CD. Each application in Argo CD is a pointer to a Helm chart that installs and configures a specific project component. When bootstrapping the deployment, a Helm values file is used to tell the application manager which components should be deployed and how they should be configured. Using this pattern gives us a significant amount of flexibility when tailoring deployments to specific environments.

Reviewer: Do we want to link directory to the `applicaton-manager` chart?

#### Disabling Components

> If you want to disable the deployment/management of certain components (for example, if you want to bring your own cluster that has ODF already installed), each set `disabled: true` in the chart's values file. For example, to disable ODF store the following file in `local/disable-odf.yaml`:

Reviewer: ```suggestion
If you want to disable the deployment/management of certain components (for example, if you want to bring your own cluster that has ODF already installed), set `disabled: true` in the chart's values file. For example, to disable ODF, create the following file in `local/disable-odf.yaml`:
```

```yaml
# Dynamically Generated Charts
application-manager:
  charts:
    # Top Level RFE App of App Chart
    rfe-automation:
      values:
        charts:
          # Cluster Configuration App of App Chart
          cluster-configs:
            values:
              charts:
                # OpenShift Data Foundations
                odf:
                  disabled: true
```

Pass this values file to helm when deploying the project. For example:

```shell
helm upgrade -i -n rfe-gitops bootstrap charts/bootstrap/ -f local/bootstrap.yaml -f examples/helm/deployment/default.yaml -f local/disable-odf.yaml
```

#### Customizing Components

> Each chart in the `charts/` directory has a default values file. These values can be overwritten using the same pattern shown above in _Disabling Components_.

Reviewer: Include a relative link to the _Disabling Components_ header

> For example, to enable processor emulation for OpenShift Virtualization set `useEmulation: true` in the chart's values file. Store the following file in `local/cnv-processor-emulation.yaml`:

Reviewer: ```suggestion
For example, to enable processor emulation for OpenShift Virtualization, set `useEmulation: true` in the chart's values file. Store the following file in `local/cnv-processor-emulation.yaml`:
```
